ShareFileJavaSDK/src/com/citrix/sharefile/api/SFApiQueryExecutor.java

```diff
@@ -209,137 +210,157 @@ private T executeQueryWithReAuthentication() throws SFServerException,
+
+                    T ret = callSuccessResponseParser(responseString);
+                    callSaveCredentialsCallback(ret);
+                    if(!SFCapabilityService.getInternal().providerCapabilitiesLoaded(urlstr)) {
+                        SFCapabilityService.getInternal().getCapabilities(urlstr,mSFApiClient);
                     }
-                    //break;
+                    return ret;
+                }
 
-                    case HttpsURLConnection.HTTP_NO_CONTENT:
-                    {
-                        return null;
-                    }
-                    //break;
+                case HttpsURLConnection.HTTP_NO_CONTENT:
+                {
+                    return null;
+                }
 
-                    case HttpsURLConnection.HTTP_UNAUTHORIZED:
-                    {
-                        Logger.d(TAG, "RESPONSE = AUTH ERROR");
+                case HttpsURLConnection.HTTP_UNAUTHORIZED:
+                {
+                    Logger.d(TAG, "RESPONSE = AUTH ERROR");
 
-                        callWipeCredentialsCallback();
+                    callWipeCredentialsCallback();
 
-                        SFFormsAuthenticationCookies formsAuthResponseCookies = SFHttpsCaller.getFormsAuthResponseCookies(url, connection, mCookieManager);
-                        if(formsAuthResponseCookies != null) {
-                            throw new SFNotAuthorizedException(SFKeywords.UN_AUTHORIZED, formsAuthResponseCookies, mReAuthContext);
-                        }
-                        return executeQueryWithReAuthentication();
+                    SFFormsAuthenticationCookies formsAuthResponseCookies = SFHttpsCaller.getFormsAuthResponseCookies(url, connection, mCookieManager);
+                    if(formsAuthResponseCookies != null) {
+                        throw new SFNotAuthorizedException(SFKeywords.UN_AUTHORIZED, formsAuthResponseCookies, mReAuthContext);
                     }
-                    //break;
-
-                    /*
-                       The ShareFile server doesn't treat 404 correctly. Eg: Try creating a
-                       duplicate folder on ShareFile. The resulting http code
-                       is ShareFile(404) vs Connectors(409).
-                       So attempt to parse the server response String for accurate error message.
-                    */
-                    case HttpsURLConnection.HTTP_NOT_FOUND:
-                    {
-                        responseString = SFHttpsCaller.readErrorResponse(connection);
-                        Logger.v(TAG, responseString);
-                        SFV3ErrorParser sfV3error = new SFV3ErrorParser(httpErrorCode, responseString, null);
-                        throw new SFNotFoundException(sfV3error.errorDisplayString());
+                    // Writelock to prevent use of token while it's being updated, note state before
+                    String tokenBefore = mSFApiClient.getOAuthToken().getAccessToken();
```

Comment:
This could also be an updated token, right? Don't we need to get `tokenBefore` somewhere in our read lock block above?

Reply:
Yes I think that's possible. I'll get the snapshot before.